core/psh/psh.c

```diff
@@ -972,7 +972,181 @@ static void psh_signalstop(int sig)
 }
 
 
-static int psh_run(void)
+static int psh_login(void)
+{
+	int insz = -1, shdwlines = 0, code;
+	const int maxinsz = 16;
+	char c, indata[maxinsz+1], indatasave[maxinsz+1];
+	char shdw[30+1];
+	struct termios mode;
+
+	tcgetattr(0, &mode);
+    mode.c_lflag &= ~(ECHO | ICANON);
+    tcsetattr(0, TCSANOW, &mode);
+
+	/* Get login from user */
+	memset(indata,'\0',16);
+	for (;;) {
+		if (insz < 0) {
+			fprintf(stdout,"Login: ");
+			fflush(0);
+			insz++;
+		}
+
+		read(STDIN_FILENO, &c, 1);
+		if (c == '\n') {
+			fprintf(stdout,"\n");
+			if (insz <= maxinsz && insz > 0) {
+				memset(indatasave,'\0',17);
+				memcpy(indatasave,indata,sizeof(char)*16);
+				memset(indata,'\0',17);
+				break;
+			}
+			else {
+				memset(indata,'\0',16);
+				insz = -1;
+				continue;
+			}
+		}
+		else if (c > 31 && c < 127) {
+			fprintf(stdout,"%c",c);
+			fflush(stdout);
+			if (insz < maxinsz)
+				indata[insz] = c;
+			insz++;
+			continue;
+		}
+		else if (c == 8 || c==127) {
+			if (insz > 0) {
+				insz--;
+				fprintf(stdout,"\b \b");
+				fflush(stdout);
+				if (insz < maxinsz)
+					indata[insz] = '\0';
+			}
+		}
+	}
+
+	/* Get password from user */
+	memset(shdw,'\0',31);
+	memset(indata,'\0',16);
+	insz = -1;
+	for (;;) {
+		if (insz < 0) {
+			printf("Password: ");
+			fflush(0);
+			insz++;
+		}
+
+		read(STDIN_FILENO, &c, 1);
+		if (c == '\n') {
+			memset(shdw,'\0',12);
+			if (insz <= maxinsz) {
+				memcpy(shdw,crypt(indata,"ABCD"),sizeof(char)*31); 
+				memset(indata,'\0',16);
+				insz = 0;
+				break;
+			}
+			else {
+				memset(indata,'\0',16);
+				insz = -1;
+				break;
+			}
+		}
+		else if (c > 31 && c < 127) {
+			if (insz < maxinsz)
+				indata[insz] = c;
+			insz++;
+		}
+		else if (c == 8) {
+			if (insz > 0) {
+				insz--;
+				if (insz < maxinsz) {
+					indata[insz] = '\0';
+				}
+			}
+		}
+	}
+	fprintf(stdout,"\n");
+	mode.c_lflag |= (ECHO | ICANON);
+	tcsetattr(0, TCSANOW, &mode);
+
+	/* Try to access /etc/passwd file */
+	FILE *shadows = NULL;
+	if (access("/etc/passwd",R_OK) == 0) {
+		shadows = fopen("/etc/passwd","r");
+		while ((c = fgetc(shadows)) != EOF) {
+			if (c=='\n')
+				shdwlines++;
+		}
+		rewind(shadows);
+		if (shdwlines == 0) {
+			fclose(shadows);
+			shadows = NULL;
+		} 
+	}
+
+	/* Compare login and shadow against /etc/passwd */
+	code = -EACCES;
+	if (shadows != NULL) {
+		for (int ii = 0; ii <= shdwlines; ii++) {
+			insz = 0;
+			memset(indata,'\0',16);
+			while ((c = fgetc(shadows)) != ':' && c > 0) {
+				if (insz < maxinsz)
+					indata[insz] = c;
+				insz++;
+			}
+			/* if login matches then check password */
+			code = -EACCES;
+			if (strcmp(indata,indatasave) ==0 && (insz <= maxinsz)) {
+				c = fgetc(shadows);
+				int gg=0;
```

Comment:
Please declare variable before the function code.

Reply:
Corrected as with previous occurrence of declaration in function code and not in the beginning of the function.